### PR TITLE
Add publish methods overloads with offset and count

### PR DESCRIPTION
**Pull request by Semax**
Hi,

This is an attempt to improve publishing methods to provide more options for serialization and array recycling usage (#194). 
I hope it'll be useful to other stan.net users.

Thanks.

### CODE REVIEW

File: src/STAN.Client/ProtocolSerializer.cs [hunk 1]
```diff
@@ -37,8 +44,9 @@ internal static byte[] createPubMsg(string clientID, string guidValue, string su
                 Subject = subject,
                 ConnID = (ByteString)connID
             };
+
             if (data != null)
```
Comment (watfordgnf) [inline src/STAN.Client/ProtocolSerializer.cs:48]:
```suggestion
            if (data != null && count > 0)
```

File: src/STAN.Client/StanConnection.cs [hunk 2]
```diff
@@ -587,21 +587,40 @@ static public string newGUID()
...
 
         public string Publish(string subject, byte[] data, EventHandler<StanAckHandlerArgs> handler)
         {
-            return publish(subject, data, handler).GUID;
+            return publish(subject, data, 0, data?.Length ?? 0, handler).GUID;
+        }
+
+        public string Publish(string subject, byte[] data, int offset, int length, EventHandler<StanAckHandlerArgs> handler)
+        {
+            return publish(subject, data, offset, length, handler).GUID;
```
Comment (watfordgnf) [inline src/STAN.Client/StanConnection.cs:605]:
```suggestion
        public string Publish(string subject, byte[] data, int offset, int count, EventHandler<StanAckHandlerArgs> handler)
        {
            return publish(subject, data, offset, count, handler).GUID;
```

**Review by watfordgnf — COMMENTED** [review]
Some minor nits/suggestions.

**Comment by Semax** [comment]
@watfordgnf, 
Thanks for your helpful suggestions.

**Review by watfordgnf — APPROVED** [review]
Pending @ColinSullivan1 and CI.

**Review by ColinSullivan1 — CHANGES_REQUESTED** [review]
This is great - thanks for the PR!

One small ask - In testing could you also sub and receive the message and check that the expected payload was received?  E.g. a new testPubSubOffset() test, similar to [testBasicPubSub()](https://github.com/nats-io/stan.net/blob/b5c310d307e1e0c78f7dd026b26bb7541785ca0a/src/Tests/IntegrationTests/BasicTests.cs#L211) except it'd only need to publish check the received payload.

**Comment by Semax** [comment]
@ColinSullivan1,
I've added the requested test.

**Review by ColinSullivan1 — APPROVED** [review]
LGTM.  Thanks!

**Pull request merged**